src/components/encoding-pane/property-editor-schema.ts

```diff
@@ -137,73 +114,83 @@ export const SEQUENTIAL_COLOR_SCHEMES = ['blues', 'greens', 'greys', 'purples',
   };
-
   const {channel} = shelfId;
-
-  if (prop === 'scale') {
-    if (nestedProp === 'type') {
-      // Filtering based on channel type & fieldDef type
+  switch (prop) {
+    case 'scale':
       const scaleTypes: string[] = getSupportedScaleTypes(shelfId, fieldDef);
-      const propertySchema = {
-        ...baseUISchema,
-        enum: scaleTypes,
-      };
-      return generateSchema(prop, nestedProp, DEFAULT_SELECT_UISCHEMA, propTab, propertySchema);
-    } else if (nestedProp === 'scheme') {
-      const propertySchema = {
-        ...baseUISchema,
-        enum: isContinuous(fieldDef) ? SEQUENTIAL_COLOR_SCHEMES : CATEGORICAL_COLOR_SCHEMES,
-      };
-      return generateSchema(prop, nestedProp, DEFAULT_SELECT_UISCHEMA, propTab, propertySchema);
-    } else if (nestedProp === 'range' || nestedProp === 'domain') {
-      return generateSchema(prop, nestedProp, isDiscrete(fieldDef) ?
-        DEFAULT_DISCRETE_SCALE_ARRAY_UISCHEMA : DEFAULT_CONTINUOUS_SCALE_ARRAY_UISCHEMA, propTab, baseUISchema);
-    }
-  } else if (prop === 'axis') {
-    if (nestedProp === 'orient') {
-      const propertySchema = {
-        ...baseUISchema,
-        enum: channel === 'y' ? ['left', 'right'] : ['top', 'bottom']
-      };
-      return generateSchema(prop, nestedProp, DEFAULT_SELECT_UISCHEMA, propTab, propertySchema);
-    } else if (nestedProp === 'title') {
-      return generateSchema(prop, nestedProp, DEFAULT_TEXT_UISCHEMA, propTab, baseUISchema);
-    }
-  } else if (prop === 'stack') {
-    const propertySchema = {
-      ...baseUISchema,
-      enum: (vlSchema as any).definitions.StackOffset.enum,
-    };
-    return generateSchema(prop, nestedProp, DEFAULT_SELECT_UISCHEMA, propTab, propertySchema);
-  } else if (prop === 'size') {
-    return generateSchema(prop, nestedProp, DEFAULT_TEXT_UISCHEMA, propTab, baseUISchema);
-  } else if (prop === 'legend') {
-    if (nestedProp === 'orient') {
-      const propertySchema = {
-        ...baseUISchema,
-        enum: (vlSchema as any).definitions.LegendOrient.enum
-      };
-      return generateSchema(prop, nestedProp, DEFAULT_SELECT_UISCHEMA, propTab, propertySchema);
-    } else if (nestedProp === 'title') {
-      return generateSchema(prop, nestedProp, DEFAULT_TEXT_UISCHEMA, propTab, baseUISchema);
-    } else if (nestedProp === 'type') {
-      const propertySchema = {
-        ...baseUISchema,
-        enum: (vlSchema as any).definitions.Legend.properties.type.enum,
-      };
-      return generateSchema(prop, nestedProp, DEFAULT_SELECT_UISCHEMA, propTab, propertySchema);
-    }
-  } else if (prop === 'format') {
-    return generateSchema(prop, nestedProp, DEFAULT_TEXT_UISCHEMA, propTab, baseUISchema);
-  } else {
-    throw new Error('Property combination not recognized');
+      return generateScaleEditorSchema(nestedProp, scaleTypes, fieldDef, baseUISchema);
+    case 'axis':
+      return generateAxisEditorSchema(nestedProp, channel as Channel, baseUISchema, propertyKey);
```

Comment:
why casting?

Reply:
Since `ShelfId` is a union type of `ShelfChannelId` and `ShelfWildcardChannelId`, and therefore `Channel` prop can be either of type `Channel` or `SHORT_WILDCARD|Wildcard<Channel>`. 

I'm not too familiar with union typing in typescript, so I figured you'd have to eventually cast it to fit the Channel type, since we are not currently supporting wildcards (ShelfWildcardChannelId)

Reply:
This issue arises quite a lot for me, as there are union types but my methods requires a specific type. How do I get around this in typescript? I know there is typeguards, but if you call the typeguard and it returns true, don't you still have to eventually cast it to that type?